src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs

```diff
@@ -67,60 +64,91 @@ private bool TryInvokeSnippetCompletion(TabKeyCommandArgs args, CompletionServic
+            //
+            //      "<line start><spaces><identifier><question><tab>"
+            //
+            // Then we take over and we show a completion list with all snippets in it
+            // in a flat list.  This enables simple browsing and filtering of all items
+            // based on what the user typed so far.
+            //
+            // If we detect this pattern, then we delete the previous character (the
+            // question mark) and we don't send the tab through to the editor.  In 
+            // essence, the <quesiton><tab> acts as the trigger, and we act as if that
+            // text never makes it into the buffer.
             Workspace workspace = null;
 
-            if (Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out workspace))
+            if (!Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out workspace))
             {
-                var documentId = workspace.GetDocumentIdInCurrentContext(subjectBuffer.AsTextContainer());
-                if (documentId != null)
-                {
-                    var document = workspace.CurrentSolution.GetDocument(documentId);
-                    if (document != null)
-                    {
-                        var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-
-                        if (caretPoint >= 2 && text[caretPoint - 1] == '?' && QuestionMarkIsPrecededByIdentifierAndWhitespace(text, caretPoint - 2, syntaxFacts))
-                        {
-                            var textChange = new TextChange(TextSpan.FromBounds(caretPoint - 1, caretPoint), string.Empty);
-                            workspace.ApplyTextChanges(documentId, textChange, CancellationToken.None);
-                            this.StartNewModelComputation(completionService, new CompletionTrigger(CompletionTriggerKind.Snippets), filterItems: false);
-                            return true;
-                        }
-                    }
-                }
+                return false;
+            }
+
+            var documentId = workspace.GetDocumentIdInCurrentContext(subjectBuffer.AsTextContainer());
+            if (documentId == null)
+            {
+                return false;
+            }
+
+            var document = workspace.CurrentSolution.GetDocument(documentId);
+            if (document == null)
+            {
+                return false;
+            }
+
+            var rules = GetCompletionService().GetRules();
+            if (rules.SnippetsRule != SnippetsRule.IncludeAfterTypingIdentifierQuestionTab)
+            {
+                return false;
             }
 
-            return false;
+            var syntaxFactsOpt = document.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFactsOpt == null ||
+                caretPoint < 2 ||
+                text[caretPoint - 1] != '?' ||
+                !QuestionMarkIsPrecededByIdentifierAndWhitespace(text, caretPoint - 1, syntaxFactsOpt))
```

Comment:
Yes.  The old code passed in the position before the question mark.  The new code passes in the position fo the question mark itself.  It made the math easier (IMO).  This was becaues the old code used to get that position, then walk backwards to the front of the line.  Now i walk from teh front of the line until i reach the question mark.